boa_engine/src/builtins/string/mod.rs

```diff
@@ -1666,101 +1669,95 @@ impl String {
+
+            // 1. Let requestedLocales be ? CanonicalizeLocaleList(locales).
+            // 2. If requestedLocales is not an empty List, then
+            //     a. Let requestedLocale be requestedLocales[0].
+            let lang = canonicalize_locale_list(args.get_or_undefined(0), context)?
+                .into_iter()
+                .next()
+                // 3. Else,
+                //     a. Let requestedLocale be ! DefaultLocale().
+                .unwrap_or_else(|| default_locale(context.icu().locale_canonicalizer()))
+                .id;
+            // 4. Let noExtensionsLocale be the String value that is requestedLocale with any Unicode locale extension sequences (6.2.1) removed.
+            // 5. Let availableLocales be a List with language tags that includes the languages for which the Unicode Character Database contains language sensitive case mappings. Implementations may add additional language tags if they support case mapping for additional locales.
+            // 6. Let locale be ! BestAvailableLocale(availableLocales, noExtensionsLocale).
+            // 7. If locale is undefined, set locale to "und".
+            let lang =
+                best_available_locale::<CaseMappingV1Marker>(lang, &context.icu().provider())
+                    .unwrap_or(LanguageIdentifier::UND);
+
+            let casemapper =
+                CaseMapping::try_new_with_locale(&context.icu().provider(), &lang.into())
+                    .map_err(|err| JsNativeError::typ().with_message(err.to_string()))?;
+
+            // 8. Let codePoints be StringToCodePoints(S).
+            let result = string.map_valid_segments(|segment| {
+                if UPPER {
+                    // 10. Else,
+                    //     a. Assert: targetCase is upper.
+                    //     b. Let newCodePoints be a List whose elements are the result of an uppercase transformation of codePoints according to an implementation-derived algorithm using locale or the Unicode Default Case Conversion algorithm.
+                    casemapper.to_full_uppercase(&segment)
+                } else {
+                    // 9. If targetCase is lower, then
+                    //     a. Let newCodePoints be a List whose elements are the result of a lowercase transformation of codePoints according to an implementation-derived algorithm using locale or the Unicode Default Case Conversion algorithm.
+                    casemapper.to_full_lowercase(&segment)
+                }
+            });
 
-            if let Some(surr) = next_unpaired_surrogate.take() {
-                upper_text.push(surr);
-            } else {
-                break;
-            }
+            // 11. Return CodePointsToString(newCodePoints).
+            Ok(result.into())
         }
 
-        // 5. Let L be ! CodePointsToString(upperText).
-        // 6. Return L.
-        Ok(js_string!(upper_text).into())
+        #[cfg(not(feature = "intl"))]
```

Comment:
Should this function even exist if intl is not active?

Reply:
Yes, the spec suggests that:

>An ECMAScript implementation that includes the ECMA-402 Internationalization API must >implement this method as specified in the ECMA-402 specification. If an ECMAScript >implementation does not include the ECMA-402 API the following specification of this method >is used ...


https://tc39.es/ecma262/#sec-string.prototype.tolocalelowercase

Reply:
Yeah, it is part of the 262 spec. It's just specialized for Intl implementors.